k4MarlinWrapper/src/components/EDM4hep2Lcio.cpp

```diff
@@ -0,0 +1,272 @@
+#include "converters/EDM4hep2Lcio.h"
+
+
+DECLARE_COMPONENT(EDM4hep2LcioTool);
+
+// Add EDM4hep to LCIO converted tracks to vector
+void EDM4hep2LcioTool::addLCIOConvertedTracks(
+  std::vector<std::pair<lcio::TrackImpl*, edm4hep::Track>>& lcio_tracks_vec,
+  const std::string& name,
+  const std::string& lcio_collection_name,
+  lcio::LCEventImpl* lcio_event)
+{
+  DataHandle<edm4hep::TrackCollection> tracks_handle {
+    name, Gaudi::DataHandle::Reader, this};
+  const auto tracks_coll = tracks_handle.get();
+
+  auto* tracks = new lcio::LCCollectionVec(lcio::LCIO::TRACK);
+
+  // Loop over EDM4hep tracks converting them to lcio tracks
+  for (int i = 0; i < tracks_coll->size(); ++i) {
+
+    const edm4hep::Track edm_tr = (*tracks_coll)[i];
+
+    auto* lcio_tr = new lcio::TrackImpl();
+
+    lcio_tr->setTypeBit( edm_tr.getType() );
+    lcio_tr->setChi2( edm_tr.getChi2() );
+    lcio_tr->setNdf( edm_tr.getNdf() );
+    lcio_tr->setdEdx( edm_tr.getDEdx() );
+    lcio_tr->setdEdxError( edm_tr.getDEdxError() );
+    lcio_tr->setRadiusOfInnermostHit( edm_tr.getRadiusOfInnermostHit() );
+    // FIXME, correctly assign hitnumbers
+    lcio_tr->subdetectorHitNumbers().resize(50);
+    for(int i =0;i<50;++i) {
+      lcio_tr->subdetectorHitNumbers()[i] = 0 ;
+    }
+
+    // Loop over the track states in the track
+    const podio::RelationRange<edm4hep::TrackState> edm_track_states = edm_tr.getTrackStates();
+    for (const auto& tr_state : edm_track_states) {
+
+      std::array<float, 15> cov = {}; // zero-init
+      cov[0]  = tr_state.covMatrix[0];
+      cov[2]  = tr_state.covMatrix[5];
+      cov[5]  = tr_state.covMatrix[9];
+      cov[9]  = tr_state.covMatrix[12];
+      cov[14] = tr_state.covMatrix[14];
+      std::array<float, 3> refP = {
```

Comment:
Why do we give only the diagonal terms? The fact that in k4DelphesSim we store only them is a something we are aware and will fix

Reply:
Because that is what was present before in your converter.
Also we need to sort this out https://github.com/key4hep/k4LCIOReader/issues/8
and this
https://github.com/key4hep/k4SimDelphes/blob/8561d3fc39b2accc99c62a561090421fad3f1e8e/converter/src/DelphesEDM4HepConverter.cc#L488-L493

Reply:
ok, I see thanks. With changing upper to lower it will make it more clear than placing individual element of the matrix

Reply:
I implemented it this way as I took it from the python example, but I will store all the elements transposed.

Reply:
Following https://github.com/key4hep/EDM4hep/issues/107 there should be no transposition

Reply:
I changed it to store the full matrix, no transpose. This produces some (many) `helix variance is negative: -X`, but that will change once the changes from EDM4hep are included.